platform/linux-generic/odp_queue_basic.c

```diff
@@ -515,6 +699,197 @@ static odp_queue_t queue_lookup(const char *name)
+		if (odp_unlikely(evv == ODP_EVENT_VECTOR_INVALID)) {
+			aggr->num_events--;
+			odp_ticketlock_unlock(&aggr->lock);
+			return num_enq;
+		}
+
+		odp_event_vector_tbl(evv, &evv_tbl);
+		for (uint16_t i = 0; i < aggr->num_events; i++)
+			evv_tbl[i] = aggr->event_tbl[i];
+
+		odp_event_vector_size_set(evv, aggr->num_events);
+		odp_event_vector_type_set(evv, aggr->event_type);
+
+		if (odp_unlikely(base_queue->enqueue(base_queue->handle,
+						     (_odp_event_hdr_t *)(uintptr_t)evv) != 0)) {
+			aggr->num_events--;
+			odp_ticketlock_unlock(&aggr->lock);
+			odp_event_vector_free(evv);
+			return num_enq;
+		}
+		aggr->num_events = 0;
+	}
+
+	odp_ticketlock_unlock(&aggr->lock);
+
+	return num_enq;
+}
+
+static int event_aggr_enq(odp_queue_t aggr_handle, _odp_event_hdr_t *event_hdr)
+{
+	queue_entry_t *aggr_queue = qentry_from_handle(aggr_handle);
+	int ret;
+
+	/* Make sure events end up in correct order within a vector */
+	if (_odp_sched_fn->ord_enq_multi(aggr_handle, (void **)&event_hdr, 1, &ret))
+		return ret == 1 ? 0 : -1;
+
+	ret = _odp_event_aggr_enq(aggr_queue, &event_hdr, 1);
+	return ret == 1 ? 0 : -1;
+}
+
+static int event_aggr_enq_multi(odp_queue_t aggr_handle, _odp_event_hdr_t *event_hdr[], int num)
+{
+	queue_entry_t *aggr_queue = qentry_from_handle(aggr_handle);
+	int ret;
+
+	/* Make sure events end up in correct order within a vector */
+	if (_odp_sched_fn->ord_enq_multi(aggr_handle, (void **)event_hdr, num, &ret))
+		return ret;
+
+	return _odp_event_aggr_enq(aggr_queue, event_hdr, num);
+}
+
+/* Called inside aggregator locks */
+static int event_aggr_enq_pending(event_aggr_t *aggr)
+{
+	odp_event_t *evv_tbl;
+	odp_event_vector_t evv;
+	queue_entry_t *base_queue = qentry_from_handle(aggr->base_queue);
+
+	if (aggr->num_events == 1) {
+		if (odp_likely(base_queue->enqueue(base_queue->handle,
+						   _odp_event_hdr(aggr->event_tbl[0])) == 0)) {
+			aggr->num_events = 0;
+			return 1;
+		}
+		return 0;
+	}
+
+	evv = odp_event_vector_alloc(aggr->pool);
+	if (odp_unlikely(evv == ODP_EVENT_VECTOR_INVALID))
+		return 0;
+
+	odp_event_vector_tbl(evv, &evv_tbl);
+	for (uint16_t i = 0; i < aggr->num_events; i++)
+		evv_tbl[i] = aggr->event_tbl[i];
+
+	odp_event_vector_size_set(evv, aggr->num_events);
+	odp_event_vector_type_set(evv, aggr->event_type);
+
+	if (odp_unlikely(base_queue->enqueue(base_queue->handle,
+					     (_odp_event_hdr_t *)(uintptr_t)evv)) != 0) {
+		odp_event_vector_free(evv);
+		return 0;
+	}
+	aggr->num_events = 0;
+	return 1;
+}
+
+static int queue_api_enq_aggr(odp_queue_t handle, odp_event_t ev,
+			      const odp_aggr_enq_param_t *param)
+{
+	queue_entry_t *queue = qentry_from_handle(handle);
+	event_aggr_t *aggr = &queue->aggr;
+
+	if (odp_unlikely(queue->type != ODP_QUEUE_TYPE_AGGR))
+		return queue->enqueue(handle, (_odp_event_hdr_t *)(uintptr_t)ev);
+
+	_ODP_ASSERT(param != NULL);
+
+	/* Cannot stash events as ordered_stash_release() ignores odp_aggr_enq_param_t */
+	_odp_sched_fn->ord_wait(handle);
```

Comment:
If there are already events in the stash, these new events will overtake them, mixing the ordering, right? Previous enqueues might have used the normal enqueue function that then stashed the events.

Reply:
Good catch, fixed in v5.